sql/src/main/java/io/druid/sql/calcite/rel/DruidQueryBuilder.java

```diff
@@ -590,65 +588,23 @@ public SelectQuery toSelectQuery(
       return null;
     }
 
-    final Filtration filtration = Filtration.create(filter).optimize(sourceRowSignature);
-    final boolean descending;
-    final int threshold;
-
-    if (limitSpec != null) {
-      // Safe to assume limitSpec has zero or one entry; DruidSelectSortRule wouldn't push in anything else.
-      if (limitSpec.getColumns().size() > 0) {
-        final OrderByColumnSpec orderBy = Iterables.getOnlyElement(limitSpec.getColumns());
-        if (!orderBy.getDimension().equals(Column.TIME_COLUMN_NAME)) {
-          throw new ISE("WTF?! Got select with non-time orderBy[%s]", orderBy);
-        }
-        descending = orderBy.getDirection() == OrderByColumnSpec.Direction.DESCENDING;
-      } else {
-        descending = false;
-      }
-
-      threshold = limitSpec.getLimit();
-    } else {
-      descending = false;
-      threshold = 0;
-    }
-
-    // We need to ask for dummy columns to prevent Select from returning all of them.
-    String dummyColumn = "dummy";
-    while (sourceRowSignature.getColumnType(dummyColumn) != null
-           || getRowOrder().contains(dummyColumn)) {
-      dummyColumn = dummyColumn + "_";
+    if (getRowOrder().isEmpty()) {
+      // Should never do a scan query without any columns that we're interested in. This is probably a planner bug.
+      throw new ISE("WTF?! Attempting to convert to Scan query without any columns?");
     }
 
-    final List<String> metrics = new ArrayList<>();
-
-    if (selectProjection != null) {
-      metrics.addAll(selectProjection.getDirectColumns());
-      metrics.addAll(selectProjection.getVirtualColumns()
-                                     .stream()
-                                     .map(VirtualColumn::getOutputName)
-                                     .collect(Collectors.toList()));
-    } else {
-      // No projection, rowOrder should reference direct columns.
-      metrics.addAll(getRowOrder());
-    }
-
-    if (metrics.isEmpty()) {
-      metrics.add(dummyColumn);
-    }
-
-    // Not used for actual queries (will be replaced by QueryMaker) but the threshold is important for the planner.
-    final PagingSpec pagingSpec = new PagingSpec(null, threshold);
+    final Filtration filtration = Filtration.create(filter).optimize(sourceRowSignature);
 
-    return new SelectQuery(
+    return new ScanQuery(
         dataSource,
         filtration.getQuerySegmentSpec(),
-        descending,
+        selectProjection != null ? VirtualColumns.create(selectProjection.getVirtualColumns()) : VirtualColumns.EMPTY,
+        ScanQuery.RESULT_FORMAT_COMPACTED_LIST,
+        0,
+        limitSpec == null || limitSpec.getLimit() == Integer.MAX_VALUE ? 0 : limitSpec.getLimit(),
```

Comment:
I think we shouldn't make this change in this patch, since there's code elsewhere that assumes `Integer.MAX_VALUE` means "no limit". Also Calcite treats limit as int in a few places anyway. So some more care would be needed to migrate that. I'll add a comment though.